Add retry logic to data ingestion

zz1874 opened the pull request:
This PR adds retry logic to the ingest_nix_graph function for handling failures during vertex and edge ingestion. It allows for multiple attempts to ingest failed items. 

Key changes include:
- Introduced a retries variable to specify the number of retry attempts.
- Implemented retry logic for vertex ingestion, logging errors and retrying failed vertices.
- Implemented retry logic for edge ingestion, logging errors and retrying failed edges.

These changes enhance the function's reliability by retrying failed ingestion attempts, ensuring successful ingestion of all items. The pull request also includes minor adjustments for improved logging and progress bar display.

Fixes #90

dorranh commented:
By the way, I know this is still a draft, but I figured it might be worth mentioning that I used a library called `backoff` to help with retries (e.g. [here](https://github.com/tweag/nixpkgs-graph-explorer/blob/4c888ee3428042f2904f12e060dce84c035480c9/api/explorer/api/gremlin.py#L102)). Not sure if it fits this use case or not but figured I would mention in case it helps: https://pypi.org/project/backoff/

zz1874 commented:
> By the way, I know this is still a draft, but I figured it might be worth mentioning that I used a library called `backoff` to help with retries (e.g. [here](https://github.com/tweag/nixpkgs-graph-explorer/blob/4c888ee3428042f2904f12e060dce84c035480c9/api/explorer/api/gremlin.py#L102)). Not sure if it fits this use case or not but figured I would mention in case it helps: https://pypi.org/project/backoff/

Thanks for mentioning this `backoff` library @dorranh ! Will definitely take a look!

smelc commented:
@zz1874> can you rebase on `main`? The current state makes it harder to read this PR's diff (e.g. I can't squash all commits to see all changes at once).

GuillaumeDesforges commented on api/explorer/api/ingest.py:
```diff
@@ -210,30 +210,61 @@ def write_edge(input: Tuple[graph.Package, graph.Edge, graph.Package]):
+            failed_packages = [
+                packages[futures.index(f)] for f in futures if f.exception()
+            ]
+            failed_packages_copy = failed_packages.copy()
```
Why do you need a copy? You've already created a new list above.

GuillaumeDesforges commented on api/explorer/api/ingest.py:
```diff
@@ -210,30 +210,61 @@ def write_edge(input: Tuple[graph.Package, graph.Edge, graph.Package]):
+            failed_packages_copy = failed_packages.copy()
+            if failed_packages_copy:
+                click.echo("\nRetrying to write nodes...")
+                for i in range(len(failed_packages_copy)):
```
Why not use the natural iterator?

```suggestion
                for failed_package in failed_packages:
```

GuillaumeDesforges commented on api/explorer/api/ingest.py:
```diff
@@ -210,30 +210,61 @@ def write_edge(input: Tuple[graph.Package, graph.Edge, graph.Package]):
+                    try:
+                        graph.insert_unique_vertex(failed_packages_copy[i], g_factory())
+                        failed_packages.remove(failed_packages_copy[i])
+                    except Exception as e:
+                        logger.exception(f"An exception occurred: {e}")
```
```suggestion
                        logger.error(f"Synchronous insertion failed for node: {failed_package_copy[i]}")
                        logger.exception(e)
```

https://docs.python.org/3/library/logging.html
> If exc_info does not evaluate as false, it causes exception information to be added to the logging message. If an exception tuple (in the format returned by [sys.exc_info()](https://docs.python.org/3/library/sys.html#sys.exc_info)) or an exception instance is provided, it is used; otherwise, [sys.exc_info()](https://docs.python.org/3/library/sys.html#sys.exc_info) is called to get the exception information.

GuillaumeDesforges commented on api/explorer/api/ingest.py:
```diff
@@ -210,30 +210,61 @@ def write_edge(input: Tuple[graph.Package, graph.Edge, graph.Package]):
+                for i in range(len(failed_packages_copy)):
+                    try:
+                        graph.insert_unique_vertex(failed_packages_copy[i], g_factory())
+                        failed_packages.remove(failed_packages_copy[i])
```
If you don't want to juggle with indicies because of that, you can build another set and later compute the difference.
https://docs.python.org/3/library/stdtypes.html#frozenset.difference

zz1874 commented:
Hi @GuillaumeDesforges and @smelc :wave: 

I have made some updates to the PR. Firstly, I rebased the branch onto the main branch to ensure a clean and up-to-date codebase. Additionally, I have made a modification to the progress bar implementation. Instead of printing out a progress bar in every `ingest_nix_package`, I have moved it to the main loop. This change will prevent unnecessary duplication of progress bars and provide a smoother user experience.

Please review the changes and let me know if you have any feedback or suggestions.
